homeassistant/components/kodi/__init__.py

```diff
@@ -2,88 +2,92 @@
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+
+from .const import (
+    CONF_WS_PORT,
+    DATA_CONNECTION,
+    DATA_KODI,
+    DATA_REMOVE_LISTENER,
+    DATA_VERSION,
+    DOMAIN,
 )
 
-SERVICE_TO_METHOD = {
-    SERVICE_ADD_MEDIA: {
-        "method": "async_add_media_to_playlist",
-        "schema": KODI_ADD_MEDIA_SCHEMA,
-    },
-    SERVICE_CALL_METHOD: {
-        "method": "async_call_method",
-        "schema": KODI_CALL_METHOD_SCHEMA,
-    },
-}
+PLATFORMS = ["media_player"]
 
 
 async def async_setup(hass, config):
     """Set up the Kodi integration."""
-    if any((CONF_PLATFORM, DOMAIN) in cfg.items() for cfg in config.get(MP_DOMAIN, [])):
-        # Register the Kodi media_player services
-        async def async_service_handler(service):
-            """Map services to methods on MediaPlayerEntity."""
-            method = SERVICE_TO_METHOD.get(service.service)
-            if not method:
-                return
-
-            params = {
-                key: value for key, value in service.data.items() if key != "entity_id"
-            }
-            entity_ids = service.data.get("entity_id")
-            if entity_ids:
-                target_players = [
-                    player
-                    for player in hass.data[DOMAIN].values()
-                    if player.entity_id in entity_ids
-                ]
-            else:
-                target_players = hass.data[DOMAIN].values()
-
-            update_tasks = []
-            for player in target_players:
-                await getattr(player, method["method"])(**params)
-
-            for player in target_players:
-                if player.should_poll:
-                    update_coro = player.async_update_ha_state(True)
-                    update_tasks.append(update_coro)
-
-            if update_tasks:
-                await asyncio.wait(update_tasks)
-
-        for service in SERVICE_TO_METHOD:
-            schema = SERVICE_TO_METHOD[service]["schema"]
-            hass.services.async_register(
-                DOMAIN, service, async_service_handler, schema=schema
-            )
-
-    # Return boolean to indicate that initialization was successful.
+    hass.data.setdefault(DOMAIN, {})
     return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Kodi from a config entry."""
+    conn = get_kodi_connection(
+        entry.data[CONF_HOST],
+        entry.data[CONF_PORT],
+        entry.data[CONF_WS_PORT],
+        entry.data[CONF_USERNAME],
+        entry.data[CONF_PASSWORD],
+        entry.data[CONF_SSL],
+        session=async_get_clientsession(hass),
+    )
+    try:
+        await conn.connect()
+        kodi = Kodi(conn)
+        await kodi.ping()
+        raw_version = (await kodi.get_application_properties(["version"]))["version"]
+    except (CannotConnectError, InvalidAuthError) as error:
+        raise ConfigEntryNotReady from error
```

Comment:
We shouldn't raise `ConfigEntryNotReady` here if the authentication fails as that isn't expected to be a recoverable error.

Reply:
Technically in Kodi `InvalidAuthError` is recoverable but it's an edge case, so yes, returning `False` makes more sense here.

(It's recoverable in case the WS port, which does not require authentication is unavailable and you fallback on the HTTP port, which requires authentication. Once the WS port is available again, it'll start working)